src/rust/engine/nailgun/src/lib.rs

```diff
@@ -111,53 +111,89 @@ impl Server {
     executor: Executor,
     config: nails::Config<N>,
     should_exit: oneshot::Receiver<()>,
-    state: watch::Sender<ServerState>,
+    exited: oneshot::Sender<Result<(), String>>,
     listener: TcpListener,
   ) {
     let exit_result = Self::accept_loop(executor, config, should_exit, listener).await;
     info!("Server exiting with {:?}", exit_result);
-    let _ = state.broadcast(ServerState::Exited(exit_result));
+    let _ = exited.send(exit_result);
   }
 
   async fn accept_loop<N: Nail>(
     executor: Executor,
     config: nails::Config<N>,
-    _should_exit: oneshot::Receiver<()>,
+    mut should_exit: oneshot::Receiver<()>,
     mut listener: TcpListener,
   ) -> Result<(), String> {
-    loop {
-      let tcp_stream = match listener.accept().await {
-        Ok((tcp_stream, _addr)) => tcp_stream,
-        Err(e) => {
+    // While connections are ongoing, they acquire `read`; before shutting down, the server
+    // acquires `write`.
+    let ongoing_connections = Arc::new(RwLock::new(()));
+
+    let result = loop {
+      let tcp_stream = match future::select(listener.accept().boxed(), should_exit).await {
+        future::Either::Left((Ok((tcp_stream, _addr)), s_e)) => {
+          // Got a connection.
+          should_exit = s_e;
```

Comment:
This mut re-assignment to preserve the should_exit channel for use in the next round reads odd. I assume something more straightforward prohibited by the need for a static lifetime.

Reply:
Yea, this is because the future moves into the `select`, and then needs to move back out.